types/block.go

```diff
@@ -917,6 +918,32 @@ func (commit *Commit) ValidateBasic() error {
 	return nil
 }
 
+// MedianTime computes the median time for a Commit based on the associated validator set.
+// The median time is the weighted median of the Timestamp fields of the commit votes,
+// with heights defined by the validator's voting powers.
+// The BFT Time algorithm ensures that the computed median time is always picked among
+// the timestamps produced by honest processes, i.e., faulty processes cannot arbitrarily
+// increase or decrease the median time.
+// See: https://github.com/cometbft/cometbft/blob/main/spec/consensus/bft-time.md
+func (commit *Commit) MedianTime(validators *ValidatorSet) time.Time {
+	weightedTimes := make([]*cmttime.WeightedTime, len(commit.Signatures))
+	totalVotingPower := int64(0)
+
+	for i, commitSig := range commit.Signatures {
+		if commitSig.BlockIDFlag == BlockIDFlagAbsent {
+			continue
+		}
+		_, validator := validators.GetByAddress(commitSig.ValidatorAddress)
+		// If there's no condition, TestValidateBlockCommit panics; not needed normally.
+		if validator != nil {
+			totalVotingPower += validator.VotingPower
```

Comment:
Why do we only consider the `totalVotingPower` of the validators whose  `commitSig.BlockIDFlag != BlockIDFlagAbsent`, shouldn't the calculation be relative to total voting power of all validators ? Like `return cmttime.WeightedMedian(weightedTimes, validators.totalVotingPower)` ? 

I was trying to find in the spec for BFT Time where it says it's the total voting power of the non absent validator votes but it was not clear to me.

Reply:
This is another good question, but the algorithm does not consider that we have all voting powers.

Moreover, the assumption here is that the `Commit` have at least `2/3 + 1` of the total voting power (otherwise, it is not a commit). Since the voting power of Byzantine validators is at most `1/3` of the total voting power, the median time is always taken from a correct processes. This because a 2/3+ quorum always have a majority of voting power from correct processes.

Reply:
The implementation related question here is why we need to provide the total voting power to this method. It is always the sum of the `Weight`s of the non-`nil` entries, therefore it could be computed inside the method.

Reply:
I see two things in the comments above:

* A question on how `totalVotingPower` is calculated. We can discuss what's better (or rather what _would have been_ better). But the fact is that we can't change the algorithm, as we might be breaking the "block format": anyone using the algorithm to validate block header would be broken. And existing chains would have some blocks some according to BFT Time v1, and other according to BFT Time v2. Moreover, we _are_ changing the algorithm: to PBTS. So this will become legacy code pretty soon.
* The need for providing `totalVotingPower` to `WeightedMedian` function. If you folks see an obvious improvement there (I didn't even think of that... I just moved the function around) please go ahead and try to improve. But bear in mind that, strictly speaking, this is not PBTS work, so it would likely have to be done on `main` directly.